### ISSUE TITLE
Are you having any troubles with table colors?

### ISSUE THREAD

**Issue by PySimpleGUI** [pyIDM/PyIDM#128]
Hiya

There is an ongoing discussion about table color problems.  I notice you're not seeming to have those.

I stick with 3.6 for my overall Python version as it's mega-stable.  The latest 3.8.3 that I have seems to have this table color problem as well.

What Python version are you running?

This little bit of code should display a table with alternating red and yellow rows, but on the latest releases of Python it has a white background.  Can you tell me what you see?

```python
import tkinter as tk
from tkinter import ttk

root = tk.Tk()
tree = ttk.Treeview()
style = ttk.Style()
style.configure('Treeview', background='red')
tree.pack()
for i in range(10):
    tree.insert("", "end", text="Item %s" % i, tag=i)
    if i % 2:
        tree.tag_configure(i, background='yellow')
root.mainloop()
```

**Comment by pyIDM** [comment]
@PySimpleGUI 
There is a known theme bug with tkinter 8.6.9 shipped with python 3.7.3+ affecting TreeView "Table",
bug url: https://core.tcl-lang.org/tk/info/509cafafae

also they proposed a bug fix:
```
def fixed_map(option):
    # Returns the style map for 'option' with any styles starting with
    # ("!disabled", "!selected", ...) filtered out

    # style.map() returns an empty list for missing options, so this should
    # be future-safe
    return [elm for elm in style.map("Treeview", query_opt=option)
            if elm[:2] != ("!disabled", "!selected")]

style.map("Treeview",
          foreground=fixed_map("foreground"),
          background=fixed_map("background")) 
```
I already tried above workaround they suggested before, but it works only if i changed the theme dynamically thru my function, but it won't work if i restarted the gui

I already ditched this fix and i am using python 3.7.1 for windows executable, however it will still affect everyone running PyPi or running from source with python 3.7.3+ 

here is your code test on my python 3.7.1:
![image](https://user-images.githubusercontent.com/58998813/85720783-0a09a800-b6f1-11ea-895f-bb504a46be10.png)


Also, as a side note I managed to change theme dynamically without having to restart the gui,

reference to change_theme() in gui.py:
https://github.com/pyIDM/PyIDM/blob/1ffd32649e08fe67d74ab2bde8ca530366192751/pyidm/gui.py#L896

**Comment by PySimpleGUI** [comment]
Wow!  Thanks for the info!

And congrats on the change_theme addition.

I've been thinking through an `apply_theme_to_window` type of mechanism.  I'm not sure how many of the update methods for the elements are lacking the settings, but I think they're mostly there.  I see you directly accessed the underlying tkinter directly rather than using the update methods.  Looks like a lot in there I'll be able to use so thanks for pointing it out!

You've hit 200 stars!  Very nice!  It's awesome to see a GUI application in Python do so well.  There are so few out there.

**Comment by PySimpleGUI** [comment]
It's so weird they closed the issue when the bug clearly remains in the latest release of Python 3.8.3.  I've tried all of the themes and it's not just the vista theme that has trouble.  Very odd and somewhat distressing to the situation as it doesn't seem like a fix is on the horizon.

**Comment by pyIDM** [comment]
> It's so weird they closed the issue when the bug clearly remains in the latest release of Python 3.8.3. I've tried all of the themes and it's not just the vista theme that has trouble. Very odd and somewhat distressing to the situation as it doesn't seem like a fix is on the horizon.

i gave up on this issue and just ignored it, when i saw tkinter developers didn’t care to fix their bug and just proposed a lousy fix

**Comment by PySimpleGUI** [comment]
It looks like the bug was fixed, it's a matter of having the Python people pick it up and use it.

It's "fixed" in 8.6.10

BUT, the [release notes for 3.9.0 Beta 3](https://docs.python.org/3.9/whatsnew/changelog.html#python-3-9-0-beta-3) does not mention tk version 8.6.10, but does specifically state that 8.6.9 is used.

![image](https://user-images.githubusercontent.com/46163555/85747796-973f0380-b6d5-11ea-8ac1-adddba9afc87.png)

It's terrible that a fix is out there, known, and it not being picked up.  I'm not sure who to even mention something to.  I got back a message right away from the tk people indicating the fix was completed, but I'm not sure where to go next to somehow get the 3.9 people to include it in upcoming releases.

Herein lies one of the many problems with OpenSource.... accountability.  There is none in this kind of situation.  In a company, this kind of thing wouldn't last long before being squished.

By the way, I added a horizontal separator!  I finally figured out how to do it.  It's not released yet to PyPI.  I'm starting to get enough features in there that I need to make a release very soon.  The features are piling up too high and need to get out there.

**Comment by pyIDM** [comment]
Good news they fixed this issue in 8.6.10, maybe it’s a matter of time until python people include it with next python releases

> By the way, I added a horizontal separator! I finally figured out how to do it. It's not released yet to PyPI. I'm starting to get enough features in there that I need to make a release very soon. The features are piling up too high and need to get out there.

Great job, keep up the good work

**Comment by PySimpleGUI** [comment]
I managed to put a "fix" into version 4.24.0, only to have to turn it off by default in 4.25.0.  You can enable it by calling:
```python
sg.set_options(enable_treeview_869_patch=True)
```

The problem with the color fix in there is that the table row highlight color gets lost.

This issue describes it:

https://github.com/PySimpleGUI/PySimpleGUI/issues/3097

Here you see the comparison of 3.6 on the left and the patched 3.8 on the right

![image](https://user-images.githubusercontent.com/46163555/86786794-66928e80-c032-11ea-986a-c0768adbcaf9.png)


What I've not been able to figure out is what ttk style setting controls this row highlight.  Do you happen to know what the setting is so that maybe I can restore it?

**Comment by PySimpleGUI** [comment]
I found the bug entry for Python where there's a struggle to get the 8.6.10 version of tk included with 3.9 Python.

https://bugs.python.org/issue39107

This seems exactly like the same kind of problem that the Mac users experienced with buttons lacking color for many years.  The tk folks fixed it, but now the Python developers seem to be struggling to get it into a release vehicle.  In the meantime, all tkinter users have lost use of color for their tables.  It's a CRAZY situation and the amount of care seems to be little.

**Comment by pyIDM** [comment]
I'm running on Manjaro Archlinux with Python version: 3.8.3 and tkinter version: 8.6.10
I didn't installed tkinter myself
it has been installed by Manjaro auto update  
I will switch to windows for testing

> I found the bug entry for Python where there's a struggle to get the 8.6.10 version of tk included with 3.9 Python.
https://bugs.python.org/issue39107

nice link, i believe they will get 8.6.10 soon

**Comment by PySimpleGUI** [comment]
I hope they test it.  Sounds like it might break my no_titlebar code, which would REALLY suck as my desktop widgets and quite a few other programs rely on that feature.

It's odd that 3.8.3 picked it up for you.  I'll check out the Windows release notes to see if maybe they did a 3.8 update.  The last 3.8.3 install I did installed tk 8.6.9.

I had to turn off the patch (make it selectable using set_options) because it messed up the selected line display.  I have not been able to figure out how to turn back on displaying the selected line.   It's displaying now as a dashed line on the Tree and not at all on Tables.

**Comment by pyIDM** [comment]
Done a lot of tests and here is my findings
1- you don't have to apply the fix to the table style itself, you just create a new style with any name 
2- you have to call the fix 2 times

I added below code to PySimpleGUI.py file at line 7065 right after `StartupTK(self)` and it did work correctly and shows the selected row highlight normally

```
def apply_theme_fix():
    # fix for table colors in tkinter 8.6.9, call style.map twice to work properly
    style = ttk.Style()
    def fixed_map(option):
        return [elm for elm in style.map('Treeview', query_opt=option) if elm[:2] != ("!disabled", "!selected")]
    style.map('Treeview', foreground=fixed_map("foreground"), background=fixed_map("background"))
    style.map('Treeview', foreground=fixed_map("foreground"), background=fixed_map("background"))

apply_theme_fix()
```

this is also an example which works fine with current PySimpleGUI without the fix


```
import PySimpleGUI as sg

theme_name = 'Reds'
sg.theme(theme_name)

headings = ['i', 'name', 'size']
col_widths = [6, 30, 10]
values = [[1, 'first', 500], [2, 'second', 200], [3, 'third', 400]]

layout = [[sg.Table(values=values, headings=headings, num_rows=9, justification='left', auto_size_columns=False,
                      vertical_scroll_only=False, key='table', enable_events=False, font='any 9',
                      max_col_width=100, col_widths=col_widths,
                      row_height=22,
                      )],
        [sg.B('Apply Theme Fix')]
]

window = sg.Window('', layout, finalize=True)

def apply_theme_fix():
    # fix for table colors in tkinter 8.6.9, call style.map twice to work properly
    style = sg.ttk.Style()
    def fixed_map(option):
        return [elm for elm in style.map('Treeview', query_opt=option) if elm[:2] != ("!disabled", "!selected")]
    style.map('Treeview', foreground=fixed_map("foreground"), background=fixed_map("background"))
    style.map('Treeview', foreground=fixed_map("foreground"), background=fixed_map("background"))


# theme fix can be called any time after window finalize, try to uncomment below or use the button later
# apply_theme_fix()


while True:
    event, values = window.read()
    if event is None:
        window.close()
        break
    elif event == 'Apply Theme Fix':
        apply_theme_fix()

```

**Comment by pyIDM** [comment]
for some reasons the above suggestion didn't work well with PyIDM, the highlight row selection is missing from the table
not sure why?
maybe because table is nested in Notebook

**Comment by PySimpleGUI** [comment]
This is what I was describing any why I had to turn enabling the patch into an Option.  Rather than defaulting to it being enabled, I default to it being offed.

Until I can figure out how to get highlighting back again, I have to leave it off.

I do see some dashed lines around tree entries, but nothing at all for tables.  The difference between 4.23.0 and 4.24.0 is making this patch controlled via an option.

If you figure out how to get the row highlight colors working I'm definitely interested!

**Comment by PySimpleGUI** [comment]
I'm a little confused why you're applying these fixes when the same technique was released in the PySimpleGUI code already a couple of releases ago. 

You'll find in my Table and Tree code (in the pack function that you were modifying), this code was added:

```python
                if tclversion_detailed == '8.6.9' and ENABLE_TREEVIEW_869_PATCH:
                    print('*** tk version 8.6.9 detected.... patching ttk treeview code ***')
                    table_style.map(style_name,
                                    foreground=_fixed_map(table_style, style_name, 'foreground'),
                                    background=_fixed_map(table_style, style_name, 'background'),)
```


My function is called _fixed_map and looks like this:

```python

    def _fixed_map(style, style_name, option):
        # Fix for setting text colour for Tkinter 8.6.9
        # From: https://core.tcl.tk/tk/info/509cafafae
        #
        # Returns the style map for 'option' with any styles starting with
        # ('!disabled', '!selected', ...) filtered out.

        # style.map() returns an empty list for missing options, so this
        # should be future-safe.
        return [elm for elm in style.map(style_name, query_opt=option) if
                elm[:2] != ('!disabled', '!selected')]
```

As I explained earlier and is in the release notes, I set the default state to not use these patches because of the very problem you ran into, the lack of row highlights.  I've not found the fix for this.  If  this can be fixed then I'll go back to having the patch on by default.

**Comment by pyIDM** [comment]
> I'm a little confused why you're applying these fixes when the same technique was released in the PySimpleGUI code already a couple of releases ago.

I believe you didn't get my point, 

try using this example with your fix on using `sg.set_options(enable_treeview_869_patch=True)`

```
import PySimpleGUI as sg
sg.set_options(enable_treeview_869_patch=True)
theme_name = 'Reds'
sg.theme(theme_name)

headings = ['i', 'name', 'size']
col_widths = [6, 30, 10]
values = [[1, 'first', 500], [2, 'second', 200], [3, 'third', 400]]

layout = [[sg.Table(values=values, headings=headings, num_rows=9, justification='left', auto_size_columns=False,
                      vertical_scroll_only=False, key='table', enable_events=False, font='any 9',
                      max_col_width=100, col_widths=col_widths,
                      row_height=22,
                      )],
]

window = sg.Window('', layout, finalize=True)
window.read()

```

the result as you said "missing highlight color for selected row

but if you add the fix here it will work properly

```
import PySimpleGUI as sg

theme_name = 'Reds'
sg.theme(theme_name)

headings = ['i', 'name', 'size']
col_widths = [6, 30, 10]
values = [[1, 'first', 500], [2, 'second', 200], [3, 'third', 400]]

layout = [[sg.Table(values=values, headings=headings, num_rows=9, justification='left', auto_size_columns=False,
                      vertical_scroll_only=False, key='table', enable_events=False, font='any 9',
                      max_col_width=100, col_widths=col_widths,
                      row_height=22,
                      )],
]

window = sg.Window('', layout, finalize=True)

def apply_theme_fix():
    # fix for table colors in tkinter 8.6.9, call style.map twice to work properly
    style = sg.ttk.Style()
    def fixed_map(option):
        return [elm for elm in style.map('Treeview', query_opt=option) if elm[:2] != ("!disabled", "!selected")]
    style.map('Treeview', foreground=fixed_map("foreground"), background=fixed_map("background"))
    style.map('Treeview', foreground=fixed_map("foreground"), background=fixed_map("background"))


# theme fix can be called any time after window finalize
apply_theme_fix()
window.read()
```

the same effect can be achieved when i modified PySimpleGUI.py file at line 7065 right after StartupTK(self) and it did work correctly and shows the selected row highlight normally
i added below code, which is slightly different than your patch i.e. calling style.map twice and put the code after showing the window
```
def apply_theme_fix():
    # fix for table colors in tkinter 8.6.9, call style.map twice to work properly
    style = ttk.Style()
    def fixed_map(option):
        return [elm for elm in style.map('Treeview', query_opt=option) if elm[:2] != ("!disabled", "!selected")]
    style.map('Treeview', foreground=fixed_map("foreground"), background=fixed_map("background"))
    style.map('Treeview', foreground=fixed_map("foreground"), background=fixed_map("background"))

apply_theme_fix()
```
just give it a try and see yourself

**Comment by PySimpleGUI** [comment]
Oh, I'm sorry... I wasn't paying close enough attention.

I'll try things out.

It's been a long week, but no excuse.  Thanks for your help.

**Comment by pyIDM** [comment]
Finally Found the fix, apparently the 'selected' state get lost somewhere 

you can modify  your `_fixed_map` function in pysimplegui.py file to the following:

```
def _fixed_map(style, style_name, option):
        # Fix for setting text colour for Tkinter 8.6.9
        # From: https://core.tcl.tk/tk/info/509cafafae
        #
        # Returns the style map for 'option' with any styles starting with
        # ('!disabled', '!selected', ...) filtered out.

        # style.map() returns an empty list for missing options, so this
        # should be future-safe.
        new_map =  [elm for elm in style.map(style_name, query_opt=option) if
                elm[:2] != ('!disabled', '!selected')]
      
        if option == 'background':
            new_map.append(('selected', 'SystemHighlight'))

        return new_map

```

tested with PyIDM and it works fine

**Comment by PySimpleGUI** [comment]
Oh wow!  I'll check this out. Thank you so much for the help.

**Comment by PySimpleGUI** [comment]
Works GREAT!

It's interesting that the row highlight color changed from a darker grayish-blue to more of a blue

## 3.8.3

![image](https://user-images.githubusercontent.com/46163555/87223480-5be63b00-c34b-11ea-9a32-a14cccc0abab.png)

## 3.6.2

![image](https://user-images.githubusercontent.com/46163555/87223486-6f91a180-c34b-11ea-990c-1190c5585df6.png)

**Comment by PySimpleGUI** [comment]
Ahh.... I see now that the line you added is what sets that color.

I've never seen this color setting discussed properly.  People were changing it by catching when an item is selected, then manually changing the color, etc.

When I changed your addition to this:
```python
        if option == 'background':
            new_map.append(('selected', 'red'))
```

Then the highlight color became red.

This is a handy thing to know because I could perhaps have a setting now for the selected color should someone want to change it.  This is the place I would add it.  I'll have to see how to change it "normally" if I add that option.

Thank you again for this help.  I can't begin to tell you how much it's helped this week.  It's not been a fun one.

**Comment by pyIDM** [comment]
This is great

> I've never seen this color setting discussed properly. People were changing it by catching when an item is selected, then manually changing the color, etc.

i didn't see this option also when searching thru stackoverflow or find a proper reference in tkinter
what i did is just printing the default style map options
for example if you execute this code :

```
import tkinter as tk
from tkinter import ttk
root = tk.Tk()

style = ttk.Style()
print(style.map('Treeview', 'background'))

root.mainloop()
```

it will give you something like this:
`[('disabled', '#d9d9d9'), ('selected', '#4a6984')]`  

or this on tkinter 8.6.9 as far i remember
`[('disabled', '#d9d9d9'), ('selected', '#4a6984'), ('!disabled', '!selected', '#FFFFFF')]`

with these flags included ('!disabled', '!selected', '#FFFFFF') it will override any other options, because it simply means if the widget is not disabled and not selected make background white
then what the fix suggested is to remove this entry from the map and keep anything else

**Comment by pyIDM** [comment]
speaking of the suggested fix 
i believeve  `new_map =  [elm for elm in style.map(style_name, query_opt=option) if  elm[:2] != ('!disabled', '!selected')]` will fail if you have something like `( '!selected', '#FFFFFF') ` or `('!selected', '!disabled', '#FFFFFF') `

this is what i use to filter out any "bang" option
`new_map = [elm for elm in style.map('Treeview', query_opt=option) if not ("!disabled" in elm[0] or "!selected" in elm[0])]`

**Comment by PySimpleGUI** [comment]
I've added a new parameter to set the selection color.  I also made a constant `TREEVIEW_OLD_SELECTED_ROW_COLOR` which is defined as:

```python
TREEVIEW_OLD_SELECTED_ROW_COLOR = '#4A6984'
```
I'm able to set it to this color or the fixed default of `SystemHighlight`.

I don't know what the normal value is for this background, but I found on both my Mint Linux version and my Python 3.6.3 version, it this color of blue:

On 3.8.3 with this old color, the new code produces this window:

![image](https://user-images.githubusercontent.com/46163555/87228261-14bd7180-c36e-11ea-8912-bb3a8bfa6435.png)

If I use the color `'SystemHighlight`' then it looks like this:

![image](https://user-images.githubusercontent.com/46163555/87228286-3dde0200-c36e-11ea-9f33-5347eb8c4644.png)


It's a different color of blue.  I'm wondering if this going to be the new color for 8.6.10 or how it's handled as I've never run the fixed version.  

What colors do you see in your code?  Did you see any difference between the 3.7.1 you were running and the color you get with the fixes?

I don't know what other user's selection colors look like.

**Comment by PySimpleGUI** [comment]
The new parameter made this table

This is the layout used in this particular demo:

```python
layout = [[sg.Table(values=data[1:][:], headings=headings, max_col_width=25,
                    # background_color='light blue',
                    auto_size_columns=True,
                    display_row_numbers=True,
                    justification='right',
                    num_rows=20,
                    alternating_row_color='lightyellow',
                    key='-TABLE-',
                    selected_row_color='green',
                    # row_height=35,
                    tooltip='This is a table')],
          [sg.Button('Read'), sg.Button('Double'), sg.Button('Change Colors')],
          [sg.Text('Read = read which rows are selected')],
          [sg.Text('Double = double the amount of data in the table')],
          [sg.Text('Change Colors = Changes the colors of rows 8 and 9')]]
```



![image](https://user-images.githubusercontent.com/46163555/87228361-bcd33a80-c36e-11ea-8ae7-aff9906193e3.png)


I've tried a few theme based colors too.  I think the best option of the theme based colors is the button's background.  So, if a user wants a theme based select color, it could look something like this:

Sometimes it works out great.. .sometimes not.  Light Blue 2
![image](https://user-images.githubusercontent.com/46163555/87228407-0f145b80-c36f-11ea-88a7-95d01b031911.png)

Dark Green 11

![image](https://user-images.githubusercontent.com/46163555/87228422-22272b80-c36f-11ea-80f9-a756e565cf4c.png)

Light Green 3

![image](https://user-images.githubusercontent.com/46163555/87228434-3539fb80-c36f-11ea-86f4-dd8227975459.png)


Those are just some random ones I tossed at it.

It's defined with this defintion:

```python
layout = [[sg.Table(values=data[1:][:], headings=headings, max_col_width=25,
                    # background_color='light blue',
                    auto_size_columns=True,
                    display_row_numbers=True,
                    justification='right',
                    num_rows=20,
                    alternating_row_color='lightyellow',
                    key='-TABLE-',
                    selected_row_color=sg.theme_button_color()[1],
                    # row_height=35,
                    tooltip='This is a table')],
```

**Comment by pyIDM** [comment]
> TREEVIEW_OLD_SELECTED_ROW_COLOR = '#4A6984'

this is the same color i have on linux python 3.8 tkinter 8.6.10
i believe i've got  'SystemHighlight' color on windows but not sure which version of python 

adding new parameter `selected_row_color` is good addition you made

> I've tried a few theme based colors too. I think the best option of the theme based colors is the button's background. So, if a user wants a theme based select color, it could look something like this:

what about "systemDefault"  theme? I think the selection won't be ok if it matched button bg

**Comment by PySimpleGUI** [comment]
I made a constant for the old highlight color that you can set when making a Table or Tree that you can set using this parameter:

```python
                   selected_row_color=sg.TREEVIEW_OLD_SELECTED_ROW_COLOR,
```

But the default is using the code you provided `'SystemHighlight'`

The example above of using the button color I was only showing as a way a user could "theme" their highlight color using existing calls and without hard coding.

I don't know how well this new blue color will work out as I've never seen this constant before.  I've only seen the other color of blue.   I read somewhere that this color is dependent upon a setting in the "Control Panel" which is kinda odd.

I found the color definitions in the 3.6 and 3.8 releases.  For 3.6 with the Vista theme, I get the same SystemHighlight color

Here is both 3.6 and 3.8 with the ttk theme set to 'vista'

![image](https://user-images.githubusercontent.com/46163555/87245605-94058080-c414-11ea-84e0-95bd1aaae15b.png)

When looking through the 3.6 folder I did find the hardcoded color #4A6984 is used in "defaults", "clam" and "alt".  And I also found the SystemHighlight in the vista among other themes.

Here's the portion of the defaults.tcl file that has this constant:
![image](https://user-images.githubusercontent.com/46163555/87245693-65d47080-c415-11ea-85d0-3ecfb92d4c21.png)



There is one difference that I see between the patched version and the 3.6 one and that's the color of the text is reversed in the 3.6. Notice it's white text in the 3.6 and left as dark text in 3.8.  It's not a hug problem in my opinion as compared to have no highlights at all or not having any colors.  It's meant to be a temporary patch.

**Comment by PySimpleGUI** [comment]
It looks like in order to really do it properly, I'm going to have to supply the ability to set both the foreground and background of the selected row.

I was able to set the text color to #FFFFFF and mimic the same behavior as before where the text is white on a blue background.

It'll take some work today but I'll make the selected row color from a single color to a tuple so that the foreground and background can be set.  This would also allow the use of theme color much easier.  You could simply set the highlight to be the same as the button colors and thus be guaranteed they will match and be visible.  I may end up going this way depending on how it all turns out.  But first the basic capabilities to set them have to be added.

To add the white text, I just changed the _fixed_map code to add the extra option for foreground:

```python
    if option == 'background':
        new_map.append(('selected', highlight_color if highlight_color is not None else 'SystemHighlight'))
    elif option == 'foreground':
        new_map.append(('selected', '#FFFFFF'))
```

**Comment by pyIDM** [comment]
I decided to track why "selected" tag got lost from our custom style
adding some print statement to _fixed_map() in PySimpleGUI.py file and i got interesting results:
`print(style.map(style_name, query_opt=option))` output is an empty list, that's why our new_map also is empty list, then we added 
"selected" tag later 

then i tried to get the default "Treeview" style map using below print statements
```
print(style.map("Treeview", 'foreground'))
print(style.map("Treeview", 'background'))
```

results:
Default foreground : `[('disabled', '#a3a3a3'), ('!disabled', '!selected', 'black'), ('selected', '#ffffff')] `
Default background :` [('disabled', '#d9d9d9'), ('!disabled !selected', '#ffffff'), ('selected', '#4a6984')] `
note `('!disabled !selected', '#ffffff') `it will be failed to get filtered by _fixed_map because it is a 2 element tuple with '!disabled !selected' as one item not 2 comma separated items

to cut the long talk:
you can get the default "selected" colors for background and foreground from default Treeview style
here is a better "I think" _fixed_map

```
def _fixed_map(style, style_name, option):
    default_map = [elm for elm in style.map("Treeview", query_opt=option) if '!' not in elm[0]]
    custom_map =  [elm for elm in style.map(style_name, query_opt=option) if '!' not in elm[0]]
    new_map = custom_map + default_map
    return new_map
```

this way you can override the default style map if you already mapped some options in your custom style before you call _fixed_map()

for example if the resulting new map = [('selected', 'green'),  ('selected', 'SystemHighlight')] , green will be chosen, because it comes first, thats why we wrote custom_map before default_map in `new_map = custom_map + default_map`

hope you get my point as i am not good in explaining things

**Comment by PySimpleGUI** [comment]
Can you try the version I've checked in that uses the theme based colors?  It's the direction I've decided to take.

The version is 4.24.0.11

Are you saying that the version of _fixed_map I'm using now will have problems?

I just found a problem with the syntax of what's checked in, but due to the way my defaults work, it shouldn't be a problem.

I like having the color of the selection match the theme.

**Comment by PySimpleGUI** [comment]
I see what you're saying about the if in the list comprehension. 

Did you encounter a problem with the existing code?  I'm not familiar enough with all this styling to fully grasp the details.

**Comment by PySimpleGUI** [comment]
I also don't want to use the blanket "Treeview" hardcoded name in the style.  I would rather it be mapped to that individual element's style setting.  Otherwise it will apply system wide and there may be situations where a user doesn't want the colors to be applied across every Tree and table.  What if they want a their table and trees to look differently?  I've run into all kinds of problems by not creating these custom styles.

**Comment by PySimpleGUI** [comment]
Oh, I see, you're using Treeview to pick up the system-wide default and then applying on top of that.  hmm.....

I lose my ability to set the custom colors in the version you posted above, right?

**Comment by PySimpleGUI** [comment]
Added back in the highlight colors. 

```python
def _fixed_map(style, style_name, option, highlight_colors=(None, None)):
    # Fix for setting text colour for Tkinter 8.6.9
    # From: https://core.tcl.tk/tk/info/509cafafae

    default_map = [elm for elm in style.map("Treeview", query_opt=option) if '!' not in elm[0]]
    custom_map = [elm for elm in style.map(style_name, query_opt=option) if '!' not in elm[0]]

    if option == 'background':
        custom_map.append(('selected', highlight_colors[1] if highlight_colors[1] is not None else ALTERNATE_TABLE_AND_TREE_SELECTED_ROW_COLORS[1]))
    elif option == 'foreground':
        custom_map.append(('selected', highlight_colors[0] if highlight_colors[0] is not None else ALTERNATE_TABLE_AND_TREE_SELECTED_ROW_COLORS[0]))

    new_map = custom_map + default_map
    return new_map
```

**Comment by PySimpleGUI** [comment]
Here's a Light Green 6 theme.

![image](https://user-images.githubusercontent.com/46163555/87249252-02a20880-c42c-11ea-81c9-6f52b6689d10.png)


The hard coded light yellow alternating row color doesn't always work, but that's a user code mess-up rather than a PySimpleGUI logic problem.

The table is defined as:

```python
layout = [[sg.Table(values=data[1:][:], headings=headings, max_col_width=25,
                    auto_size_columns=True,
                    display_row_numbers=True,
                    justification='right',
                    num_rows=20,
                    alternating_row_color='lightyellow',
                    key='-TABLE-',
                    selected_row_colors='red on yellow',
                    tooltip='This is a table')]
```

I definitely like the theme being used as the color selection by default.  The manually selected colors follow the new button format of a tuple or a string with "on" in it.
